src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2240,7 +2240,22 @@ void CodeGen::genSSE42Intrinsic(GenTreeHWIntrinsic* node, insOpts instOptions)
         {
             assert(instOptions == INS_OPTS_NONE);
 
-            assert(!op2->isUsedFromReg() || (op2->GetRegNum() != targetReg) || (op1Reg == targetReg));
+#ifdef DEBUG
+            if (op2->isUsedFromReg() && (op2->GetRegNum() == targetReg) && (op1Reg != targetReg))
+            {
+                // The move we are issuing is going to overwrite 'op2' with
+                // 'op1'. This is still ok if 'op1' and 'op2' are the same
+                // local; normally they would be in the same register in that
+                // case, but LSRA may have inserted a copy for one of the
+                // operands (either due to stress or for any other reason).
+                // This check mirrors what happens in genCodeForBinary.
+                GenTree* op1Skip = op1->gtSkipReloadOrCopy();
+                GenTree* op2Skip = op2->gtSkipReloadOrCopy();
+                assert(op1Skip->OperIsLocalRead() && GenTree::Compare(op1Skip, op2Skip) &&
+                       "'mov' to set up crc32 RMW is going to overwrite op2!");
+            }
```

Comment:
> This is a fairly verbose check that seemingly we'd need to be doing everywhere we currently have assert(!op2->isUsedFromReg() || (op2->GetRegNum() != targetReg) || (op1Reg == targetReg)); or similar (for both x86/x64 and Arm64).
> 
> Should it be extracted to a helper method?
> 
> -- This is seemingly also an issue everywhere we're doing such a check (which I think is before most RMW moves). Is there something unique to crc32 that's causing it to trigger here and not in the other places?

Can you point to some of them? If they use this pattern and LSRA builds the uses in the same way, then yeah they can likely hit this problem.

I think a preferable fix would be what I pointed out: skip the LSRA optimization that currently happens inside `AddDelayFreeUses`, since that optimization really is not correct (in the sense that it does not properly model the constraints that we have here). We will probably need to keep it for the normal RMW ops due to diffs, but for other cases like crc32 the optimization does not seem important enough to warrant.

Reply:
We have various asserts in the `emitIns_SIMD_*` helpers: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/emitxarch.cpp#L8837. I believe we also have similar being done for Arm64 and in the various `*codegen*.cpp` files (`codegenxarch` and `hwintrinsiccodegenarm64` for example).

I don't think we want to unnecessarily pessimize the codegen here by forcing them to never match either. We really just want a basic check that the RMW handling (i.e. the `mov tgt, op1`) isn't overwriting one of the other input operands (in other words, that they were correctly marked delay free).

Reply:
I extracted a function to check the "same local" case. It wasn't straightforward for me to find the other places that need this check, so let's just cross that bridge if we see them in the future. I think that's ok since the assert is harmless in this case.